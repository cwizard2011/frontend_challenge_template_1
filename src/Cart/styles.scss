@import '../../public/colors.scss';

.tag-button {
  background-color: $grey;
  border-radius: 50%;
  width: 25px;
  height: 25px;
  text-align: center;
  margin-right: 10px;
}

.cart-checkout-button {
  border-radius: 20px;
  color: #FFF;
  background-color: #f7436b;
  border-color: #f7436b;
  box-shadow: 0.5rem 0.5rem 3rem rgba(0,0,0,0.2);
  width: 150px;
  height: 40px;
  cursor: pointer;
  font-family: 'Montserrat';
  outline: none !important;
}

.cart-checkout-button-disabled {
  border-radius: 20px;
  color: #FFF;
  background-color: darkgray;
  border-color: darkgray;
  box-shadow: 0.5rem 0.5rem 3rem rgba(0,0,0,0.2);
  max-height: 100px;
  max-width: 300px;
  padding-right: 10px;
  padding-left: 10px;
  padding-top: 10px;
  padding-bottom: 10px;
  cursor: not-allowed;
  font-family: 'Montserrat';
}

.item {
  margin-top: -1px;
}

.product-count {
  border-radius: 20px;
  border: 1px solid $grey;
  width: 40px;
  height: 25px;
  text-align: center;
  margin-right: 10px;
}

.product-section {
  display: 'flex';
  flex-basis: 25%;
  width: 400px;
  justify-content: space-between;
}

.back-to-shop {
  width: 150px;
  height: 40px;
  border-radius: 20px;
  background-color: #FFF;
  color: black;
  border-color: #FFF;
  box-shadow: 0.5rem 0.5rem 3rem rgba(0,0,0,0.2);
  font-family: 'Montserrat';
  outline: none !important;
  color: #f7436b;
}

.product-in-cart-wrapper {
  display: flex;
  justify-content: space-around;
  flex-direction: row;
  margin-top: 25px;
  margin-bottom: 25px;

  & p {
    margin-top: 20px;
    margin-left: 30px;
    font-size: 20px;
    font-family: 'Montserrat';
  }

  & button {
    width: 150px;
    height: 40px;
    border-radius: 20px;
    color: #f7436b;
    border-color: #FFF;
    box-shadow: 0.5rem 0.5rem 3rem rgba(0,0,0,0.2);
    margin-top: 15px;
    font-family: 'Montserrat';
    outline: none !important;
  }
}

.decrement {
  width: 20px;
  height: 20px;
  background-color: #D3D3D3;
  border-radius: 50%;
  text-align: center;
  margin-right: 10px;
  cursor: pointer;
  font-family: 'Montserrat';

  & p {
    margin-top: -3px;
  }
}

.increment {
  width: 20px;
  height: 20px;
  background-color: #D3D3D3;
  border-radius: 50%;
  text-align: center;
  margin-left: 10px;
  cursor: pointer;
  font-family: 'Montserrat';

  & p {
    margin-top: -2px;
  }
}

.header-styles {
  font-family: 'Montserrat';
}

.login-to-checkout {
  color: #f7436b;
  margin-top: 10px;
  cursor: pointer;
}
