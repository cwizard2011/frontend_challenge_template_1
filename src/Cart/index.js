import React, { Component } from 'react';
import toastr from 'toastr';
import axios from 'axios';
import { getCartItems, getCartPrice } from '../utils/apiCalls';
import { withToastManager } from 'react-toast-notifications';
import Navbar from '../Navbar';
import { Table } from 'reactstrap';
import Spinner from '../Spinner';
import './styles.scss';

export class Cart extends Component {
  state = {
    productIncart: [],
    total_price: '',
    cartID: '',
    isLoading: false,
    isLoggedIn: false
  }

  async componentDidMount() {
    this.setState({ isLoading: true })
    const getCartID = await localStorage.getItem('cartId');
    const userID = await localStorage.getItem('user');
    if (!userID) {
      this.setState({ isLoggedIn: false })
    } else {
      this.setState({ isLoggedIn: true })
    }
    if (!getCartID) {
      this.setState({ productIncart: [], isLoading: false })
    } else {
      const cartItem = await getCartItems(getCartID);
      const totalPrice = await getCartPrice(getCartID);
      this.setState({ productIncart: cartItem.data, total_price: totalPrice.data.total_amount, isLoading: false });
    }
  }

  changeQuantityCount = async (change, itemID) => {
    const { productIncart } = this.state;
    const getCartID = await localStorage.getItem('cartId');
    const foundItem = productIncart.find(product => product.item_id === itemID);
    let { item_id, quantity, image, product_id } = foundItem;
    let newQty;
    if (change === 'add') {
      newQty = quantity + 1;
    } else if (change === 'remove') {
      newQty = quantity - 1;
    }
    const data = { quantity: newQty };
    const returnedResponse = await axios.put(`https://backendapi.turing.com/shoppingcart/update/${item_id}`, data);
    const totalPrice = await axios.get(`https://backendapi.turing.com/shoppingcart/totalAmount/${getCartID}`);
    const modifiedProductInCart = returnedResponse.data.find(product => product.item_id === item_id);
    const mergedData = { ...modifiedProductInCart, image, product_id };
    const indexOfReplacedElement = productIncart.indexOf(foundItem);
    productIncart[indexOfReplacedElement] = mergedData;
    this.setState({ productIncart, total_price: totalPrice.data.total_amount })
  }

  placeOrder = async() => {
    const getCartID = await localStorage.getItem('cartId');
    const userData = await localStorage.getItem('userData');
    const accessToken = await localStorage.getItem('user');
    const parsedAccessToken = JSON.parse(accessToken);
    const parsedUser = JSON.parse(userData);
    const { shipping_region_id } = parsedUser;
    const data = { cart_id: getCartID, shipping_id: shipping_region_id, tax_id: 2 };
    try {
      const url = 'https://backendapi.turing.com/orders';
      const options = {
        method: 'POST',
        data,
        url,
        headers: {
          'USER-KEY': `${parsedAccessToken.accessToken}`,
          'Content-type': 'application/json'
        }
      }
      const postOrder = await axios(options);
      const cartItem = await axios.get(`https://backendapi.turing.com/shoppingcart/${getCartID}`);
      const totalPrice = await axios.get(`https://backendapi.turing.com/shoppingcart/totalAmount/${getCartID}`)
      this.setState({ productIncart: cartItem.data, total_price: totalPrice.data.total_amount});
    } catch (error) {
      toastr.error('An error occurred.')
    }
  }

  removeProductFromCart = async(itemID) => {
    const getCartID = await localStorage.getItem('cartId');
    await axios.delete(`https://backendapi.turing.com/shoppingcart/removeProduct/${itemID}`);
    const cartItem = await axios.get(`https://backendapi.turing.com/shoppingcart/${getCartID}`);
    const totalPrice = await axios.get(`https://backendapi.turing.com/shoppingcart/totalAmount/${getCartID}`);
    this.setState({ productIncart: cartItem.data, total_price: totalPrice.data.total_amount })
  }

  clearCart = async() => {
    const getCartID = await localStorage.getItem('cartId');
    const response = await axios.delete(`https://backendapi.turing.com/shoppingcart/empty/${getCartID}`);
    this.setState({ productIncart: response.data, total_price: '0.00' })
  }

  render() {
    const { productIncart, total_price, isLoading, isLoggedIn } = this.state;
    const { history } = this.props;
    if (isLoading) {
      return <Spinner />
    }

    return (
      <div id="cart">
        <Navbar
          productIncart={productIncart}
          cartPrice={total_price}
          history={history}
        />
        {productIncart.length === 0 && (
          <div style={{ textAlign: 'center', marginTop: 150 }}>
            <p style={{ fontSize: 25 }}>You have no items in cart</p>
            <button className="back-to-shop"
            onClick={() => history.push('/')}
          >
              Back to Shop
            </button>
          </div>
          )}
          {productIncart.length > 0 &&
          (
            <div className="product-in-cart-wrapper">
              <p>
                {productIncart.length} item(s) in cart
              </p>
              <button
                onClick={this.clearCart}
                >
                Clear Cart
              </button>
            </div>
          )
          }
        {productIncart.length > 0 && <Table>
        <thead>
          <tr>
            <th className="header-styles">Item</th>
            <th className="header-styles">Size</th>
            <th className="header-styles">Color</th>
            <th className="header-styles">Quantity</th>
            <th className="header-styles">Price</th>
          </tr>
        </thead>
        <tbody>
        {
          productIncart.map((product, index) => (
            <tr key={index}>
              <td>
              {product.image &&
              <div style={{ display: 'flex' }}>
              <img
                  src={require(`../assets/product_images/${product.image}`)}
                  alt=""
                  style={{
                    height: 80,
                    width: 80
                  }}
                />
                <div style={{ marginLeft: 15 }}>
                  <p className="cart-item-title" style={{ fontFamily: 'Montserrat' }}>{product.name}</p>
                  <i
                    className="fas fa-times cart-item-remove" style={{ color: 'red', marginRight: 10, cursor: 'pointer' }}
                    onClick={() => this.removeProductFromCart(product.item_id)}
                  >
                  </i>
                  <span style={{ fontFamily: 'Montserrat', fontSize: 14 }}>Remove</span>
                </div>
                </div>
                }
              </td>
              <td className="cart-item-size" style={{ fontFamily: 'Montserrat' }}>{product.attributes.split(',')[0] || 'N/A'}</td>
              <td className="cart-item-color" style={{ fontFamily: 'Montserrat' }}>{product.attributes.split(',')[1] || 'N/A'}</td>
              <td style={{ display: 'flex' }}>
                <div className="decrement">
                  <p
                    className="decrease-quantity"
                    onClick={() => this.changeQuantityCount('remove', product.item_id)}
                  >
                    -
                  </p>
                </div>
                <p style={{
                  backgroundColor: '#FFF',
                  borderStyle: 'solid',
                  borderWidth: 'thin',
                  borderColor: '#D3D3D3',
                  borderRadius: '50%',
                  width: 30,
                  textAlign: 'center',
                  marginTop: -3,
                  fontFamily: 'Montserrat'
                }}
                className="cart-item-quantity"
                >{product.quantity}</p>
                <div className="increment">
                  <p
                    className="increase-quantity"
                    onClick={() => this.changeQuantityCount('add', product.item_id)}
                  >
                    +
                  </p>
                </div>
              </td>
              <td className="cart-item-price" style={{ fontFamily: 'Montserrat' }}>${product.subtotal}</td>
            </tr>
          ))
        }
        <tr>
          <td>
            Total
          </td>
          <td id="cartTotalPriceValue">
            ${total_price}
          </td>
        </tr>
        </tbody>
      </Table>}
        {productIncart.length > 0 && <div style={{ display: 'flex', justifyContent: 'space-around' }}>
          <button className="back-to-shop"
            onClick={() => history.push('/')}
          >
            Back to shop
          </button>
          {isLoggedIn ?
          (
            <button
              id="btnCheckout"
              className="cart-checkout-button"
              onClick={() => history.push('/checkout')}
            >
              Checkout
            </button>
          ) : <p onClick={() => history.push('/login')} className="login-to-checkout">Log in to proceed to checkout</p>}
        </div>}
      </div>
    )
  }
}

const CartWithToast = withToastManager(Cart);

export default CartWithToast;
