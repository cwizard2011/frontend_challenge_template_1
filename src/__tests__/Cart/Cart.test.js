import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Cart } from '../../Cart';

describe('<Cart />', () => {
  const props = {
    history: {
      location: {
        pathname: ''
      }
    }
  };
  const wrapper = mount(<Cart {...props} />);

  it('should mount without crashing', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
