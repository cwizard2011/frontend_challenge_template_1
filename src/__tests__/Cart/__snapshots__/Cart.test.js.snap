// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should mount without crashing 1`] = `
<Cart
  history={
    Object {
      "location": Object {
        "pathname": "",
      },
    }
  }
>
  <Spinner>
    <div
      className="spinner"
    />
  </Spinner>
</Cart>
`;
